### PR TITLE
Migrating views components to TypeScript.

### PR DESCRIPTION
**Pull request by dennisoelkers**
## Motivation and Context

This PR is migrating all remaining views components to TypeScript.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-web-interface/src/views/components/DebugOverlay.tsx [hunk 1]
```diff
@@ -49,11 +51,4 @@ const DebugOverlay = ({ currentView, searches, show, onClose }: Props) => (
   </BootstrapModalWrapper>
 );
 
-DebugOverlay.propTypes = {
-  currentView: PropTypes.object.isRequired,
-  onClose: PropTypes.func.isRequired,
-  searches: PropTypes.object.isRequired,
-  show: PropTypes.bool.isRequired,
-};
-
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/DebugOverlay.tsx:58]: Why did you remove the prop types here? The import still exists.

File: graylog2-web-interface/src/views/components/DebugOverlay.tsx [hunk 2]
```diff
@@ -26,6 +26,8 @@ import BootstrapModalWrapper from 'components/bootstrap/BootstrapModalWrapper';
 import type { ViewStoreState } from 'views/stores/ViewStore';
 import type { SearchStoreState } from 'views/stores/SearchStore';
 
+import View from '../logic/views/View';
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/DebugOverlay.tsx:29]: This import is not needed.

File: graylog2-web-interface/src/views/components/Field.tsx [hunk 3]
```diff
@@ -24,14 +24,14 @@ import CustomPropTypes from './CustomPropTypes';
...
 import InteractiveContext from './contexts/InteractiveContext';
 
-type Props = {|
-  children?: React.Node,
+type Props = {
+  children?: React.ReactNode,
   disabled?: boolean,
   name: string,
-  menuContainer: ?HTMLElement,
+  menuContainer?: HTMLElement,
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/Field.tsx:31]: Even though the prop is optional, shouldn't this be `HTMLElement | null | undefined`?

File: graylog2-web-interface/src/views/components/QueryBar.tsx [hunk 4]
```diff
@@ -55,7 +56,16 @@ const onCloseTab = (queryId, currentQuery, queries) => {
...
 };
 
-const QueryBar = ({ children, queries, queryTitles, viewMetadata }) => {
+type Props = {
+  children?: React.ReactElement,
+  queries: string[],
+  queryTitles: Immutable.Map<string, string>,
+  viewMetadata: {
+    activeQuery: string,
+  },
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/QueryBar.tsx:65]: What about `viewMetadata: ViewMetadata`?

File: graylog2-web-interface/src/views/components/widgets/FieldSelect.tsx [hunk 5]
```diff
@@ -30,9 +30,11 @@ const ValueComponent = styled.span`
 `;
 
 type Props = {
-  onChange: ({ label: string, value: string }[]) => void,
+  onChange: (newFields: { label: string, value: string }[]) => void,
   fields: Immutable.List<FieldTypeMapping>,
-  value: ?{field: string}[],
+  value: { field: string }[] | undefined | null,
+  allowOptionCreation?: boolean,
+  inputId?: string,
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/widgets/FieldSelect.tsx:37]: Looks like these two need some default props.

File: graylog2-web-interface/src/views/pages/ShowDashboardInBigDisplayMode.tsx [hunk 6]
```diff
@@ -28,24 +28,21 @@ import View from 'views/logic/views/View';
 import { RefreshActions } from 'views/stores/RefreshStore';
 import type { UntypedBigDisplayModeQuery } from 'views/components/dashboard/BigDisplayModeConfiguration';
 import withParams from 'routing/withParams';
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/pages/ShowDashboardInBigDisplayMode.tsx:30]: This import can be removed

File: graylog2-web-interface/src/views/components/sidebar/Sidebar.tsx [hunk 7]
```diff
@@ -23,22 +23,22 @@ import styled, { css } from 'styled-components';
...
 
 type Props = {
-  children: React.Element<any>,
+  children: React.ReactNode,
   queryId: string,
   results: QueryResult,
-  searchPageLayout: ?SearchPageLayout,
-  sections: Array<SidebarSection>,
+  searchPageLayout?: SearchPageLayout,
+  sections?: Array<SidebarSection>,
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/sidebar/Sidebar.tsx:41]: These props have no corresponding `defaultProps` declaration.

File: graylog2-web-interface/src/views/components/visualizations/worldmap/__tests__/WorldMapVisualization.test.tsx [hunk 8]
```diff
@@ -32,7 +34,7 @@ jest.mock('../MapVisualization', () => 'map-visualization');
 describe('WorldMapVisualization', () => {
   // $FlowFixMe: type is always defined
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/visualizations/worldmap/__tests__/WorldMapVisualization.test.tsx:35]: This `// $FlowFixMe:` can be removed

**Review by linuspahl — CHANGES_REQUESTED** [review]
Looks good to me, just found some smaller things which can be improved.

File: graylog2-web-interface/src/views/components/DebugOverlay.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/DebugOverlay.tsx:58]: We cannot get these in sync with the types of prop (`object` vs. `ViewStoreState`), so I removed them, as we do not rely on them for this debug component. I removed the import though.

File: graylog2-web-interface/src/views/components/QueryBar.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/QueryBar.tsx:65]: :white_check_mark:

File: graylog2-web-interface/src/views/components/messagelist/CustomHighlighting.tsx [hunk 9]
```diff
@@ -26,7 +26,7 @@ import PossiblyHighlight from './PossiblyHighlight';
 import Highlight from './Highlight';
 
 type Props = {
-  children: ?React.Node,
+  children: React.ReactElement,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/messagelist/CustomHighlighting.tsx:29]: :white_check_mark:

File: graylog2-web-interface/src/views/components/messagelist/HighlightMessageInQuery.tsx [hunk 10]
```diff
@@ -23,11 +23,14 @@ import type { Location } from 'routing/withLocation';
 import HighlightMessageContext from '../contexts/HighlightMessageContext';
 
 type Props = {
-  children: ?React.Node,
+  children: React.ReactElement,
   location: Location,
+  query?: any,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/messagelist/HighlightMessageInQuery.tsx:28]: :white_check_mark:  (I added it to support a test, but the prop is not used in that component it all, so I moved it to the test.)

File: graylog2-web-interface/src/views/components/widgets/FieldSelect.tsx (hunk 5 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/widgets/FieldSelect.tsx:37]: :white_check_mark:

File: graylog2-web-interface/src/views/pages/ShowDashboardInBigDisplayMode.tsx (hunk 6 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/pages/ShowDashboardInBigDisplayMode.tsx:30]: :white_check_mark:

File: graylog2-web-interface/src/views/components/sidebar/Sidebar.tsx (hunk 7 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/sidebar/Sidebar.tsx:41]: :white_check_mark:

File: graylog2-web-interface/src/views/components/visualizations/worldmap/__tests__/WorldMapVisualization.test.tsx (hunk 8 above)
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/visualizations/worldmap/__tests__/WorldMapVisualization.test.tsx:35]: :white_check_mark:

File: graylog2-web-interface/src/views/components/messagelist/MessageField.tsx [hunk 11]
```diff
@@ -91,7 +91,11 @@ MessageField.propTypes = {
   currentView: CustomPropTypes.CurrentView.isRequired,
   fieldName: PropTypes.string.isRequired,
   fieldType: CustomPropTypes.FieldType.isRequired,
-  message: PropTypes.object.isRequired,
+  message: PropTypes.exact({
+    id: PropTypes.string.isRequired,
+    index: PropTypes.string.isRequired,
+    fields: PropTypes.object.isRequired,
+  }).isRequired,
```
[the remark this answers is not in the document]
Comment (dennisoelkers) [reply graylog2-web-interface/src/views/components/messagelist/MessageField.tsx:98]: We need to use `exact` in most cases as TS's objects are always exact. I completed the definition and added a `Message` custom prop type that we can reuse.

**Review by linuspahl — APPROVED** [review]
Look good to me!

**Pull request merged**